### PR TITLE
Update `relative_url_root` docs based on @pixeltrix explanation

### LINKED ISSUE #42248: Update `relative_url_root` docs
A few things should be updated for the relative_url_root docs.

https://guides.rubyonrails.org/configuring.html#deploy-to-a-subdirectory-relative-url-root

> Firstly, regarding your point about `config.action_controller.relative_url_root` - that's what the configuration used to be called in earlier versions of Rails but has been moved to the general application config because it affects a number of areas in Rails - Action Mailer, Assets as well as Action Controller. It seems we missed one place to rename it in the guides and I apologise for that.

`config.action_controller.relative_url_root` should be changed to `config.relative_url_root`

`config.ru`:

``` ruby
map RelativeRoot::Application.config.relative_url_root || '/' do
  run Rails.application
end
```

> There's a very good reason why you need that there and it relates to the CGI environment variables [`SCRIPT_NAME`][1] and [`PATH_INFO`][2]. The former is what you'd consider the relative url root and Rails needs Rack to that set correctly when the app receives `call` with the request environment hash - if it's not set then you might as well just wrap all your routes in a `scope("/sub") { ... }` block. Using `map` creates a `Rack::URLMap` instance which [you can see from the code][3] sets `SCRIPT_NAME` and `PATH_INFO` to the desired values. When a Rails application has a mounted Engine (or other Rack app), those values then get adjusted so that you can easily mount them at different URLs without altering the Engine. The `SCRIPT_NAME` value takes precedence over the relative url config when there is a request context like in a controller but in other contexts like an Action Mailer view or an Active Job task there is no `SCRIPT_NAME` so there the relative url config is used. I hope this helps you understand the complex interactions at play here - it's not a simple matter of just prefixing `config.relative_url_root` onto every generated url.

This explanation and the `Rack::URLMap` example would be an amazing addition to the docs, especially if there was an included section for Puma.

[1]: https://datatracker.ietf.org/doc/html/rfc3875#section-4.1.13
[2]: https://datatracker.ietf.org/doc/html/rfc3875#section-4.1.5
[3]: https://github.com/rack/rack/blob/master/lib/rack/urlmap.rb

_Originally posted by @pixeltrix in https://github.com/rails/rails/issues/42243#issuecomment-842747962_

### PR DESCRIPTION
**Pull request by zzak**
https://github.com/rails/rails/issues/42243#issuecomment-842747962

Fixes #42248

### CODE REVIEW

File: guides/source/rails_on_rack.md [hunk 1]
```diff
@@ -80,6 +80,18 @@ To find out more about different `rackup` options, you can run:
 $ rackup --help
 ```
 
+#### Prefix your app using `config.relative_url_root`
+
+When using `rackup` and you are deploying your app to a subdirectory, you need to use `map` to define the root url to your app when generating the `Rack::URLMap` structure.
```
Comment (pixeltrix) [inline guides/source/rails_on_rack.md:85]: Is `rackup` correct here? Pretty sure that puma reads config.ru and possibly others?

**Review by pixeltrix — COMMENTED** [review]
Probably need a bit on how `:script_name` will override the config for a request. The statement at the top of the guide says it assumes knowledge of how Rack works so I don't think we need too much detail and certainly not a whole explanation of how CGI variables like `SCRIPT_NAME` and `PATH_INFO` work 😄

File: guides/source/rails_on_rack.md (hunk 1 above)
Comment (zzak) [reply guides/source/rails_on_rack.md:85]: @pixeltrix Good point, I was using this in the context of a rackup file, would it make more sense to just say "When using `puma` or other Ruby web servers that use  the rackup file `config.ru`..."?

Comment (zzak) [reply guides/source/rails_on_rack.md:85]: Also maybe Rails on Rack guide isn't the best place 🤔

File: guides/source/rails_on_rack.md [hunk 2]
```diff
@@ -80,6 +80,18 @@ To find out more about different `rackup` options, you can run:
...
+
+```ruby
+# Rails.root/config.ru
+
+map RelativeRoot::Application.config.relative_url_root || '/' do
```
Comment (arianf) [inline guides/source/rails_on_rack.md:90]: `RelativeRoot` might be confusing here, `TestApp` would probably be more appropriate

Comment (zzak) [reply guides/source/rails_on_rack.md:90]: @arianf Fear not! Do you have opinions re this comment: https://github.com/rails/rails/pull/42303/files#r641911853

**Pull request closed without merging**